File: bindings/java-ffm/src/test/java/org/maplibre/nativeffi/internal/LogCallbackStateTest.java
package org.maplibre.nativeffi.internal;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.maplibre.nativeffi.LogRecord;
import org.maplibre.nativeffi.LogSeverity;
import org.maplibre.nativeffi.MapLibre;
import org.maplibre.nativeffi.internal.c.MapLibreNativeC;
import org.maplibre.nativeffi.internal.c.mln_log_callback;

final class LogCallbackStateTest {
  @BeforeAll
  static void loadNativeLibrary() {
    NativeTestSupport.loadNativeLibrary();
  }

  @AfterEach
  void clearCallback() {
    MapLibre.clearLogCallback();
  }

  @Test
  void callbackStateSurvivesUntilCleared() {
    var seen = new AtomicReference<LogRecord>();
    MapLibre.setLogCallback(
        record -> {
          seen.set(record);
          return true;
        });

    var state = LogCallbackState.currentForTesting();
    assertNotNull(state);
    try (var arena = Arena.ofConfined()) {
      var message = MemoryUtil.allocateCString(arena, "hello");
      var consumed =
          mln_log_callback.invoke(
              state.stubForTesting(),
              MemorySegment.NULL,
              MapLibreNativeC.MLN_LOG_SEVERITY_INFO(),
              MapLibreNativeC.MLN_LOG_EVENT_GENERAL(),
              7,
              message);
      assertEquals(1, consumed);
      assertEquals("hello", seen.get().message());
      assertEquals(LogSeverity.INFO, seen.get().severity());
    }
  }

  @Test
  void callbackExceptionsDoNotUnwindIntoNativeCode() {
    MapLibre.setLogCallback(
        record -> {
          throw new AssertionError("boom");
        });

    var state = LogCallbackState.currentForTesting();
    assertNotNull(state);
    try (var arena = Arena.ofConfined()) {
      var message = MemoryUtil.allocateCString(arena, "ignored");
      var consumed =
          mln_log_callback.invoke(
              state.stubForTesting(),
              MemorySegment.NULL,
              MapLibreNativeC.MLN_LOG_SEVERITY_ERROR(),
              MapLibreNativeC.MLN_LOG_EVENT_GENERAL(),
              0,
              message);
      assertEquals(0, consumed);
    }
  }
}

File: bindings/kotlin-native/src/nativeTest/kotlin/org/maplibre/nativeffi/log/LogCallbackStateTest.kt
package org.maplibre.nativeffi.log

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.cstr
import kotlinx.cinterop.memScoped
import org.maplibre.nativeffi.Maplibre
import org.maplibre.nativeffi.internal.callback.LogCallbackState

@OptIn(ExperimentalForeignApi::class)
class LogCallbackStateTest {
  @Test
  fun processGlobalLogCallbackCopiesRecordsContainsExceptionsAndClearsState() {
    val records = mutableListOf<LogRecord>()
    try {
      Maplibre.setLogCallback(
        LogCallback { record ->
          records += record
          true
        }
      )
      val state = LogCallbackState.currentForTesting()
      memScoped { assertEquals(1U, state?.invoke(1U, 3U, 7L, "hello".cstr.getPointer(this))) }
      assertEquals(LogSeverity.INFO, records.single().severity)
      assertEquals(LogEvent.PARSE_STYLE, records.single().event)
      assertEquals(7L, records.single().code)
      assertEquals("hello", records.single().message)

      Maplibre.setLogCallback(LogCallback { throw IllegalStateException("contained") })
      assertEquals(0U, LogCallbackState.currentForTesting()?.invoke(2U, 6U, 0L, null))
    } finally {
      Maplibre.clearLogCallback()
      assertNull(LogCallbackState.currentForTesting())
    }
  }

  @Test
  fun logSeverityMasksRejectUnknownInputs() {
    assertEquals(1 shl 1, LogSeverity.INFO.nativeMask)
    kotlin.test.assertFailsWith<IllegalArgumentException> { LogSeverity.UNKNOWN.nativeMask }
  }
}
